### PR TITLE
ARROW-12768: [C++] Stricter signed zero comparison in tests

### PR DESCRIPTION
**Pull request by pitrou**
Add a new parameter to EqualOptions() to choose whether signed zeros should compare equal.

Ensure that said parameter is set to false in unit tests.

### CODE REVIEW

**Comment by pitrou** [comment]
cc @edponce

File: cpp/src/arrow/compare.h [hunk 1]
```diff
@@ -48,6 +48,16 @@ class EqualOptions {
     return res;
   }
 
+  /// Whether or not zeros with differing signs are considered equal.
+  bool signed_zeros_equal() const { return signed_zeros_equal_; }
```
Comment (pitrou) [inline cpp/src/arrow/compare.h:52]: Not sure about the naming here.

Comment (lidavidm) [reply cpp/src/arrow/compare.h:52]: signed_zeros_equal sounds fine to me.

**Review by lidavidm — APPROVED** [review]
LGTM.

**Comment by lidavidm** [comment]
Though, it seems there are a few more tests to adjust for Windows and MacOS.

**Review by lidavidm — APPROVED** [review]

**Comment by pitrou** [comment]
AppVeyor build (green): https://ci.appveyor.com/project/pitrou/arrow/builds/41873970

**Pull request closed without merging**